test/cri-containerd/container_update_test.go

```diff
@@ -0,0 +1,334 @@
+			requiredFeatures: []string{featureWCOWProcess},
+			runtimeHandler:   wcowProcessRuntimeHandler,
+			sandboxImage:     imageWindowsNanoserver,
+			containerImage:   imageWindowsNanoserver,
+			cmd:              []string{"cmd", "/c", "ping", "-t", "127.0.0.1"},
+		},
+		{
+			name:             "WCOW_Hypervisor",
+			requiredFeatures: []string{featureWCOWHypervisor},
+			runtimeHandler:   wcowHypervisorRuntimeHandler,
+			sandboxImage:     imageWindowsNanoserver,
+			containerImage:   imageWindowsNanoserver,
+			cmd:              []string{"cmd", "/c", "ping", "-t", "127.0.0.1"},
+		},
+		{
+			name:             "LCOW",
+			requiredFeatures: []string{featureLCOW},
+			runtimeHandler:   lcowRuntimeHandler,
+			sandboxImage:     imageLcowK8sPause,
+			containerImage:   imageLcowAlpine,
+			cmd:              []string{"top"},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			requireFeatures(t, test.requiredFeatures...)
+
+			if test.runtimeHandler == lcowRuntimeHandler {
+				pullRequiredLcowImages(t, []string{test.sandboxImage})
+			} else if test.runtimeHandler == wcowHypervisorRuntimeHandler {
+				pullRequiredImages(t, []string{test.sandboxImage})
+			}
+
+			podRequest := &runtime.RunPodSandboxRequest{
+				Config: &runtime.PodSandboxConfig{
+					Metadata: &runtime.PodSandboxMetadata{
+						Name:      t.Name(),
+						Namespace: testNamespace,
+					},
+				},
+				RuntimeHandler: test.runtimeHandler,
+			}
+
+			client := newTestRuntimeClient(t)
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			podID := runPodSandbox(t, client, ctx, podRequest)
+			defer removePodSandbox(t, client, ctx, podID)
+			defer stopPodSandbox(t, client, ctx, podID)
+
+			containerRequest := &runtime.CreateContainerRequest{
+				Config: &runtime.ContainerConfig{
+					Metadata: &runtime.ContainerMetadata{
+						Name: t.Name() + "-Container",
+					},
+					Image: &runtime.ImageSpec{
+						Image: test.containerImage,
+					},
+					Command: test.cmd,
+				},
+				PodSandboxId:  podID,
+				SandboxConfig: podRequest.Config,
+			}
+
+			containerID := createContainer(t, client, ctx, containerRequest)
+			defer removeContainer(t, client, ctx, containerID)
+
+			startContainer(t, client, ctx, containerID)
+			defer stopContainer(t, client, ctx, containerID)
+
+			// make request to increase cpu shares
+			updateReq := &runtime.UpdateContainerResourcesRequest{
+				ContainerId: containerID,
+			}
+
+			if test.runtimeHandler == lcowRuntimeHandler {
+				updateReq.Linux = &runtime.LinuxContainerResources{
+					CpuShares: 5000,
+				}
+			} else {
+				updateReq.Windows = &runtime.WindowsContainerResources{
+					CpuShares: 5000,
+				}
+			}
+
+			if _, err := client.UpdateContainerResources(ctx, updateReq); err != nil {
+				t.Fatalf("updating container resources for %s with %v", containerID, err)
+			}
+		})
+	}
+}
+
+func Test_Container_UpdateResources_CPUShare_NotRunning(t *testing.T) {
```

Comment:
How does this test work? From my understanding shim actually creates & starts a container only when a start container request is sent. So here when the update resources request is sent the container should not exist right?

Reply:
The idea for this test is that CRI supports updating resources even if the container is only in "CREATED" state. In that case, a call to update the resources would only update the container's spec and when the container is started, it would use that updated value.